### ISSUE TITLE
Obtaining hidden states of the RNN

### ISSUE THREAD

**Issue by Jihadik** [fchollet/keras#4013]
Suppose we have some trained RNN and some sample with `T` time-steps. I want to get hidden states of the network after each time-step. What is the best way of doing that?

**Details**
1. Architecture of my network is the following:
[Masking, LSTM, LSTM, TimeDistributed, Dense]
2. I want to get hidden states from the last LSTM. I'm just popping last two layers (Dense and TimeDistributed) and making `predict_on_batch` with the intention that predictions will be hidden states.

Unfortunately, something is wrong since all hidden states (obtained as was said earlier) looks like some fixed vector is multiplied by different constants. I looked at the weight matrices and they don't look like identity matrices.

**Comment by Jihadik** [comment]
Okay, I found that `predict_on_batch` or just `predict` outputs tensor of shape `(time, hidden_size)` instead of `(batch, time, hidden_size)`.
Any ideas why?

**Comment by dieuwkehupkes** [comment]
It sounds like your approach should work, this is how I usually do it as well. Can you give some small example that illustrates that it doesn't?

Why would you want the results organised by batch? You can easily retrieve that anyhow if you know your batchsize by only taking part of your output.

**Comment by Jihadik** [comment]
Model creation code:
```python
model = Sequential()
model.add(Masking(mask_value=0.0,
                  input_shape=(None, feats_dim * window)))
for dim in hidden_dims:
    model.add(LSTM(dim, return_sequences=True, consume_less='gpu'))
model.add(TimeDistributed(Dense(nclasses, activation='softmax')))
model.compile(loss=objective,
              optimizer=optimizer(learning_rate, momentum,
                                  clipping, algorithm))
```

After training this model:
```python
with open(model_path + '-model.json', 'r') as fin:
    model = model_from_json(fin.read())
model.load_weights(model_path + '-weights.hdf5')
model.pop() # Activation
model.pop() # TimeDistributed
```

Finally:
```python
preds = model.predict_on_batch(batch)
print(batch.shape)
print(preds.shape)
```
It prints:
```
(32, 924, 40)
(924, 128)
```
For me it seems that the output shape should be `(32, 924, 128)`.

**Comment by dieuwkehupkes** [comment]
It is really hard to reproduce this problem with all the variable names (that have no values). When I put some values for `feats_dim`, `window`, `hidden_dims` and `nclasses` and apply the model on some randomly generated input it does give me a three dimensional output.

One thing that might be going wrong is that you are popping one layer too much from your model. `TimeDistributed is` a wrapper, not a layer, so with the first `pop()` you are already removing it. Try `model.summary()` to see what your model looks like. But even then the output shouldn't really look like this. Please provide a small script or a gist or something that reproduces the problem.

**Comment by Jihadik** [comment]
Sorry, with shapes it was my fault (typo). What about vectors, I changed matrix to random and it started producing different outputs, hence RNN learned something trivial.
Thanks!

**Comment by mohammadrezaebrahimi** [comment]
@dieuwkehupkes, I was wondering if the described approach would provide all the hidden states emitted in each timestep. I need to store all hidden states at each timestep in a file to capture their changes during training.  In other words, I need a way to serialize the states as they changes during training? Any idea in this regard would be appreciated.

**Comment by dieuwkehupkes** [comment]
@mohammadrezaebrahimi yes I think it should give all hidden states emitted at each timestep, this is at least what I also used it for!

**Comment by mohammadrezaebrahimi** [comment]
@dieuwkehupkes thanks for your reply. Since my model is doing sequence classification (rather than sequence tagging), I noticed that I need to use return_state=true, rather than return_sequences=true (I only need the last state).So I wrote the following in my LSTM layer:
`z, hs, cs  = LSTM(100, return_state=True)(x)`

Hoping that I can get the hidden states and the cell states via hs and cs. After training the model I can see _hs_ is a tensor variable. Is there anyway I can access it via a theano function? Please let me know if i am doing this wrong. --Thanks!
